File: projects/closed_loop_servo_pid/sim/pid_controller.py
"""Python port of firmware/pid_controller.c, kept 1:1 with the C update
equation so simulation numbers reflect what the embedded code would do.
"""


def clamp(value, lo, hi):
    return max(lo, min(hi, value))


class PIDController:
    def __init__(self, kp, ki, kd, output_min, output_max):
        self.kp = kp
        self.ki = ki
        self.kd = kd
        self.integral = 0.0
        self.prev_error = 0.0
        self.has_prev_error = False
        self.output_min = output_min
        self.output_max = output_max
        self.integral_min = output_min
        self.integral_max = output_max

    def reset(self):
        self.integral = 0.0
        self.prev_error = 0.0
        self.has_prev_error = False

    def update(self, setpoint, measurement, dt):
        if dt <= 0:
            return 0.0

        error = setpoint - measurement

        self.integral += error * dt
        self.integral = clamp(self.integral, self.integral_min, self.integral_max)

        derivative = 0.0
        if self.has_prev_error:
            derivative = (error - self.prev_error) / dt
        self.prev_error = error
        self.has_prev_error = True

        output = self.kp * error + self.ki * self.integral + self.kd * derivative
        return clamp(output, self.output_min, self.output_max)

File: projects/closed_loop_servo_pid/firmware/pid_controller.c
#include "pid_controller.h"

static float clamp(float value, float lo, float hi) {
    if (value < lo) return lo;
    if (value > hi) return hi;
    return value;
}

void pid_init(PIDController *pid, float kp, float ki, float kd,
              float output_min, float output_max) {
    pid->kp = kp;
    pid->ki = ki;
    pid->kd = kd;
    pid->integral = 0.0f;
    pid->prev_error = 0.0f;
    pid->has_prev_error = 0;
    pid->output_min = output_min;
    pid->output_max = output_max;
    /* Clamp the integral term to the output range to limit windup. */
    pid->integral_min = output_min;
    pid->integral_max = output_max;
}

void pid_reset(PIDController *pid) {
    pid->integral = 0.0f;
    pid->prev_error = 0.0f;
    pid->has_prev_error = 0;
}

float pid_update(PIDController *pid, float setpoint, float measurement, float dt) {
    if (dt <= 0.0f) {
        return 0.0f;
    }

    float error = setpoint - measurement;

    pid->integral += error * dt;
    pid->integral = clamp(pid->integral, pid->integral_min, pid->integral_max);

    float derivative = 0.0f;
    if (pid->has_prev_error) {
        derivative = (error - pid->prev_error) / dt;
    }
    pid->prev_error = error;
    pid->has_prev_error = 1;

    float output = pid->kp * error + pid->ki * pid->integral + pid->kd * derivative;
    return clamp(output, pid->output_min, pid->output_max);
}
